### PR TITLE
HPCC-31951 Add extra logging and protection to CDeltaWriter

### PR DESCRIPTION
**Pull request by jakesmith**
Add guards and extra logging to the CDeltaWriter in the event it runs into trouble.

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: dali/base/dasds.cpp [hunk 1]
```diff
@@ -1377,56 +1385,67 @@ class CDeltaWriter : implements IThreaded
...
+                                pendingTransactionsSem.wait();
 
-                        writeRequested = false;
-                    }
-                    if (signalWhenAllWritten)
-                    {
-                        signalWhenAllWritten = false;
-                        allWrittenSem.signal();
+                            writeRequested = false;
+                            if (signalWhenAllWritten) // can only be true if writeRequested was true
```
Comment (jakesmith) [inline dali/base/dasds.cpp:1421]: @ghalliday: NB: I moved this inside the if (writeRequested) block, so it is clearer that is only sent and possible when writeRequested is true

File: dali/base/dasds.cpp [hunk 2]
```diff
@@ -2198,6 +2217,9 @@ void CDeltaWriter::addToQueue(CTransactionItem *item)
     }
     else // here if exceeded transationQueueLimit, transactionMaxMem or exceeded time threshold (deltaSaveThresholdSecs)
     {
+        if (isAborted()) // critical situation if delta writer is no longer running (there will have been previous errors)
```
Comment (richardkchapman) [inline dali/base/dasds.cpp:2220]: Why isAborted() rather than just testing aborted (here, but not elsewhere)?

Comment (jakesmith) [reply dali/base/dasds.cpp:2220]: no particular reason, I'll change it so consistent.

**Comment by jakesmith** [comment]
@richardkchapman - change per your comment, and rebased.

File: dali/base/dasds.cpp [hunk 3]
```diff
@@ -1377,56 +1385,67 @@ class CDeltaWriter : implements IThreaded
             threaded.join();
         }
     }
+    bool isAborted() const { return aborted; }
```
Comment (richardkchapman) [inline dali/base/dasds.cpp:1388]: I don't think this is used any more?

**Review by richardkchapman — APPROVED** [review]
One comment, but not fatal

**Review by ghalliday — APPROVED** [review]

**Pull request merged**